### ISSUE TITLE
ActiveRecord Sum (and other calculation operators) don't honor distinct attributes

### ISSUE THREAD

**Issue by SimonKaluza** [rails/rails#16791]
When using the ActiveRecord "sum" calculation with the Postgres adapter (and I believe the others though I haven't tested them), any attributes added in a "distinct" clause are not used to filter/trim the dataset.  Though "DISTINCT" is added to the SQL, it is incorrectly added _outside_ the SUM operation, which effectively ignores the relevant distinct attributes.

This seems to be directly related to issue https://github.com/rails/rails/issues/5554 and issue https://github.com/rails/rails/pull/6970.  Though the pull request related to those issues seemed to fix the undesirable behavior for the "count" ActiveRecord::Calculations.

I've written a short example script to display this behavior (there's also some Rails setup code included so it could be run outside of a Rails project, like in an IRB console easily):

```
require 'active_record'
require 'pg'


####    SETUP MIGRATIONS/DATABASE (Moved out of canonical Rails structure and into one file for demonstration)    ####
PG_SPEC = {
    :adapter  => 'postgresql',
    :host     => 'localhost',
    :database => 'my_test_db',
    :username => 'my_db_super_user',
    :encoding => 'utf8'
}

ActiveRecord::Base.establish_connection(PG_SPEC.merge('database' => 'postgres', 'schema_search_path' => 'public'))
ActiveRecord::Base.connection.drop_database PG_SPEC[:database] rescue nil
ActiveRecord::Base.connection.create_database(PG_SPEC[:database])
ActiveRecord::Base.establish_connection(PG_SPEC)

class CreateMyModels < ActiveRecord::Migration
  def self.change
    create_table :users do |t|
      t.integer :summable_attr
      t.timestamps
    end

    create_table :user_groups do |t|
      t.integer :user_id
      t.integer :group_id
      t.timestamps
    end

    create_table :groups do |t|
      t.timestamps
      t.string :status
    end
  end
end

CreateMyModels.change

####    SETUP ACTIVERECORD, POPULATE DATABASE, AND DEMONSTRATE ODD QUERY BEHAVIOR    ####

class User < ActiveRecord::Base
  has_many :user_groups
  has_many :groups, through: :user_groups

  # Should only return users that are active
  def self.all_active_users
    User.joins(:user_groups).joins(:groups).distinct(true).where('groups.status' => 'active')
  end
end

class Group < ActiveRecord::Base
  has_many :user_groups
  has_many :users, through: :user_groups
end

class UserGroup < ActiveRecord::Base
  belongs_to :group
  belongs_to :user
end

g1 = Group.create(status: 'active')
g2 = Group.create(status: 'inactive')

u1 = User.create(summable_attr: 1 )
u2 = User.create(summable_attr: 2 )
u3 = User.create(summable_attr: 3 )
u4 = User.create(summable_attr: 4 )

UserGroup.create(group_id: g1.id, user_id: u1.id)
UserGroup.create(group_id: g1.id, user_id: u2.id)
UserGroup.create(group_id: g1.id, user_id: u3.id)
UserGroup.create(group_id: g1.id, user_id: u4.id)


UserGroup.create(group_id: g2.id, user_id: u1.id)
UserGroup.create(group_id: g2.id, user_id: u2.id)
UserGroup.create(group_id: g2.id, user_id: u3.id)
UserGroup.create(group_id: g2.id, user_id: u4.id)

# Perform the sum using a Postgres AR query gives an incorrect result.
# Outputs "Total Sum: 20"
# Uses SQL: SELECT DISTINCT SUM("users"."summable_attr") AS sum_id FROM "users" INNER JOIN "user_groups" ON "user_groups"."user_id" = "users"."id" INNER JOIN "user_groups" "user_groups_users_join" ON "user_groups_users_join"."user_id" = "users"."id" INNER JOIN "groups" ON "groups"."id" = "user_groups_users_join"."group_id" WHERE "groups"."status" = 'active'
puts "Total Sum: #{User.all_active_users.sum('summable_attr')}"

# The correct summation SQL should be something like:
# Outputs "Total Sum: 10"
puts "Total Sum: #{ActiveRecord::Base.connection.execute(%q[SELECT DISTINCT SUM(DISTINCT "users"."id") FROM "users" INNER JOIN "user_groups" ON "user_groups"."user_id" = "users"."id" INNER JOIN "user_groups" "user_groups_users_join" ON "user_groups_users_join"."user_id" = "users"."id" INNER JOIN "groups" ON "groups"."id" = "user_groups_users_join"."group_id" WHERE "groups"."status" = 'active'])[0]["sum"]}"

# Perform the sum using a Ruby inject operator gives us the correct result (at quite a heavy performance/memory cost)
# Outputs "Total Sum: 10"
# Uses SQL: SELECT DISTINCT "users".* FROM "users" INNER JOIN "user_groups" ON "user_groups"."user_id" = "users"."id" INNER JOIN "user_groups" "user_groups_users_join" ON "user_groups_users_join"."user_id" = "users"."id" INNER JOIN "groups" ON "groups"."id" = "user_groups_users_join"."group_id" WHERE "groups"."status" = 'active'
puts "Total Sum: #{User.all_active_users.map {|u| u.summable_attr}.inject(:+)}"

# However, the "count" from ActiveRecord::Calculations seems to work correctly:
# Outputs "Total Count: 4"
# Uses SQL: SELECT DISTINCT COUNT(DISTINCT "users"."id") FROM "users" INNER JOIN "user_groups" ON "user_groups"."user_id" = "users"."id" INNER JOIN "user_groups" "user_groups_users_join" ON "user_groups_users_join"."user_id" = "users"."id" INNER JOIN "groups" ON "groups"."id" = "user_groups_users_join"."group_id" WHERE "groups"."status" = 'active'
puts "Total Count: #{User.all_active_users.count}"

```

**Comment by ntdb** [comment]
+1

Thanks for the temporary fix, @SimonKaluza. In my case I'm caching the calculated values to offset the performance hit.

**Comment by ClaudioFloreani** [comment]
Still there in **Rails 4.2.2**. I think this is only the tip of a more extended issue: https://github.com/rails/rails/issues/21799

**Comment by georges** [comment]
Confirm problem still exists in Rails 4.2.1. Using sum on a habtm with a uniq constraint leads to duplicates being added

**Comment by sgrif** [comment]
Thank you for the confirmations. We have tagged the issue as still present and no not need additional people to confirm at this time.

**Comment by TrevorHinesley** [comment]
For the record, a work-around for this is to change:

`Model.all.sum(:column)`

to:

`Model.all.map(&:column).inject(:+)`

The caveat is this can result in `nil` if `Model.all` is empty. Something like the following may be a viable alternative if that's an issue for you:

`Model.all.map(&:column).inject(:+) || 0`

**Comment by jeffp** [comment]
`Model.all.map(&:column).inject(:+)` is not a very pleasing work around from a performance standpoint.  The desire is to have the summation perform correctly in SQL and not have to resort to code and data structures for the summation.

**Comment by TrevorHinesley** [comment]
@jeffp you're correct, but it's about all we got at this point ;)

**Comment by flyfy1** [comment]
Encountered this issue today.. On Rails5, using Solidus. Solidus framework make have use of `sum` calculation.. therefore potentially have some bugs of making non-distinct result.

**Comment by joevandyk** [comment]
@TrevorHinesley You can also do `Model.all.lto_a.sum(&:column)`

**Comment by avit** [comment]
Also related: #7121 and https://groups.google.com/forum/?fromgroups=#!topic/rubyonrails-core/mwBehQcEduU

Count options for distinct and grouping should belong on the `count` method, not pulled from other parts of the relation.

**Comment by trcarden** [comment]
If you are looking for a less awful workaround you can use the power of AR Association Extensions (http://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#module-ActiveRecord::Associations::ClassMethods-label-Association+extensions) and redefine `sum` and whatever other calculations you need to use a subquery when needed. 

```ruby 
  has_many :things, -> { distinct }, through: :other_things do
    def sum(column_name)
      Thing.where(id: self).sum(column_name)
    end
  end
```

If you are feeling ultra fancy you can create a module and use the `extends` feature to dry it up too.

**Comment by trcarden** [comment]
Also reported here: https://github.com/rails/rails/issues/16791

**Comment by TrevorHinesley** [comment]
@trcarden thanks! Not sure I'd consider overwriting a core AR method a less awful workaround though... lol hopefully this will just be fixed soon.

**Comment by trcarden** [comment]
Haha, Yeah I can understand that feeling and by using extensions rather than a module monkey patch the change is localized to just that association so it will only affect `things.sum` and if you are ultra conservative you can always just rename it to `things.distinctly_sum` or similar to prevent complete overrides.

But I will say with that you get:
- Speed (the db can rip through a sum pretty quickly)
- Memory efficiency (not hydrating a list of models to get a simple sum)
- Local overrides (you are not changing the framework whole hog)
- It works! haha

**Comment by rafaelfranca** [comment]
Fixed by https://github.com/rails/rails/pull/30169

**Comment by CodingAnarchy** [comment]
This is supposedly fixed in Rails 5.2.0, but I am still hitting it there.

**Comment by trappist** [comment]
Still exists in 5.2.3
